File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Features/VoiceAgent/VoiceAgentMicDriver.ts
/**
 * VoiceAgentMicDriver.ts
 *
 * Audio ingress for the voice agent. The C ABI owns NO microphone access
 * (rac_voice_agent.h "Audio-Ingress Contract"): the platform SDK captures raw
 * mic frames and pushes them continuously into the C core via
 * `rac_voice_agent_feed_audio_proto` (`voiceAgentFeedAudioProto`). The core
 * performs energy-based utterance segmentation and runs the STT -> LLM -> TTS
 * turn pipeline itself, returning the synthesized reply inline for playback.
 * This driver is therefore a thin capture -> feed -> play loop with NO SDK-side
 * VAD. Turn VoiceEvents fan out through the handle callback, so
 * `RunAnywhere.streamVoiceAgent()` collectors observe them without extra wiring.
 *
 * Mirrors `sdk/runanywhere-swift/.../VoiceAgentMicDriver.swift` and
 * `sdk/runanywhere-kotlin/.../VoiceAgentMicDriver.kt`. Segmentation/endpointing
 * lives in the C core; frames captured while a turn is processing are dropped
 * by the core (and the bounded queue here) so the device's own TTS playout is
 * not re-fed (strict turn-taking, no barge-in).
 */

import { AudioCaptureManager } from '../VoiceSession/AudioCaptureManager';
import { AudioPlaybackManager } from '../VoiceSession/AudioPlaybackManager';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import { requireNativeModule } from '../../native';
import { arrayBufferToBytes } from '../../services/ProtoBytes';
import { VoiceAgentResult as VoiceAgentResultMessage } from '@runanywhere/proto-ts/voice_agent_service';
import { AudioEncoding } from '@runanywhere/proto-ts/voice_events';

const SAMPLE_RATE_HZ = 16_000;
const CHANNELS = 1;
/** Bounded backlog so a slow turn cannot grow the queue without limit. */
const CHANNEL_CAPACITY = 128;
/** Poll interval when no captured frames are pending. */
const FEED_IDLE_SLEEP_MS = 20;

/**
 * Captures mic audio and feeds raw frames to the in-core voice agent.
 * {@link start} runs until {@link stop} is called. The capture callback only
 * enqueues frames; a single async feed loop drains them and calls the core,
 * which blocks for the duration of a turn when an utterance closes and returns
 * the synthesized reply inline. We play it and drop any backlog captured during
 * the turn so the device's own playout is not re-fed.
 */
export class VoiceAgentMicDriver {
  private readonly logger = new SDKLogger('VoiceAgentMic');
  private readonly capture = new AudioCaptureManager();
  private readonly playback = new AudioPlaybackManager();

  private stopped = false;
  private queue: Uint8Array[] = [];

  /** Begin mic capture + feed loop. Resolves once capture has started. */
  async start(): Promise<void> {
    const granted = await this.capture.requestPermission();
    if (!granted) {
      throw new Error('Microphone permission denied');
    }
    this.stopped = false;
    this.queue = [];
    // The voice agent runs a single full-duplex (.playAndRecord) session for the
    // whole turn-taking loop so TTS replies can play through the speaker while
    // the mic stays live. Activate it BEFORE startRecording so capture reuses it
    // instead of forcing the output-only .record session (which silences replies
    // and trips cannotStartPlaying). Mirrors the iOS Swift driver's
    // configureVoiceAudioSession(); no-op on Android.
    await this.capture.activateAudioSession();
    await this.capture.startRecording((chunk) => this.enqueueChunk(chunk));
    this.logger.info('Voice-agent mic capture started');
    void this.feedLoop();
  }

  /** Stop capture + playback and reset state. */
  async stop(): Promise<void> {
    if (this.stopped) return;
    this.stopped = true;
    try {
      this.capture.stopRecording();
    } catch {
      /* noop */
    }
    try {
      this.playback.stop();
    } catch {
      /* noop */
    }
    this.queue = [];
    this.logger.info('Voice-agent mic capture stopped');
  }

  private enqueueChunk(chunk: Uint8Array): void {
    if (this.stopped || chunk.byteLength === 0) return;
    this.queue.push(chunk);
    if (this.queue.length > CHANNEL_CAPACITY) {
      this.queue.splice(0, this.queue.length - CHANNEL_CAPACITY);
    }
  }

  private drainChunks(): Uint8Array[] {
    if (this.queue.length === 0) return [];
    const drained = this.queue;
    this.queue = [];
    return drained;
  }

  /**
   * Drains captured frames and feeds them to the core. A non-empty
   * `synthesizedAudio` in the returned result means the core closed an
   * utterance and ran a full turn this call; we play the reply and drop any
   * backlog captured during the turn so the device's own playout is not re-fed.
   */
  private async feedLoop(): Promise<void> {
    const native = requireNativeModule();
    while (!this.stopped) {
      const chunks = this.drainChunks();
      if (chunks.length === 0) {
        await VoiceAgentMicDriver.sleep(FEED_IDLE_SLEEP_MS);
        continue;
      }

      for (const chunk of chunks) {
        if (this.stopped) return;
        try {
          const resultBytes = await native.voiceAgentFeedAudioProto(
            VoiceAgentMicDriver.toArrayBuffer(chunk),
            SAMPLE_RATE_HZ,
            CHANNELS,
            AudioEncoding.AUDIO_ENCODING_PCM_S16_LE,
            false
          );
          if (this.stopped) return;
          const bytes = arrayBufferToBytes(resultBytes);
          if (bytes.byteLength === 0) continue; // utterance still open

          const result = VoiceAgentResultMessage.decode(bytes);
          if (result.errorMessage && result.errorMessage.length > 0) {
            this.logger.warning(`Voice turn failed: ${result.errorMessage}`);
          }
          if (await this.playReply(result)) {
            // Drop frames captured during the turn + playback.
            this.queue = [];
          }
        } catch (error) {
          this.logger.warning(
            `Voice feed threw: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }
    }
  }

  /**
   * Play the turn's synthesized reply through the shared playback sink.
   * Commons returns `synthesized_audio` as a complete WAV
   * (`rac_audio_float32_to_wav`), so the bytes are handed to the native player
   * as-is — not re-encoded. Returns true when a reply was played.
   */
  private async playReply(
    result: ReturnType<typeof VoiceAgentResultMessage.decode>
  ): Promise<boolean> {
    const audio = result.synthesizedAudio;
    if (!audio || audio.byteLength === 0) return false;

    const wav = VoiceAgentMicDriver.toArrayBuffer(audio);
    this.logger.info(`Playing agent reply (${audio.byteLength} WAV bytes)`);
    try {
      await this.playback.playWav(wav);
    } catch (error) {
      this.logger.warning(
        `Agent reply playback failed: ${error instanceof Error ? error.message : String(error)}`
      );
    }
    return true;
  }

  private static sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  private static toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
    return bytes.buffer.slice(
      bytes.byteOffset,
      bytes.byteOffset + bytes.byteLength
    ) as ArrayBuffer;
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/features/VoiceAgent/Services/VoiceAgentMicDriver.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * VoiceAgentMicDriver.kt
 *
 * Audio ingress for the voice agent. The C ABI owns NO microphone access
 * (rac_voice_agent.h "Audio-Ingress Contract"): the platform SDK captures raw
 * mic frames and pushes them continuously into the C core via
 * `rac_voice_agent_feed_audio_proto`. The core performs energy-based utterance
 * segmentation and runs the STT -> LLM -> TTS turn pipeline itself, returning
 * the synthesized reply inline for playback. This driver is therefore a thin
 * capture -> feed -> play loop with NO SDK-side VAD; turn VoiceEvents fan out
 * to the handle callback, so `RunAnywhere.streamVoiceAgent()` collectors
 * observe them without extra wiring.
 */

package com.runanywhere.sdk.features.VoiceAgent.Services

import ai.runanywhere.proto.v1.AudioEncoding
import ai.runanywhere.proto.v1.VoiceAgentResult
import com.runanywhere.sdk.features.STT.Services.AudioCaptureManager
import com.runanywhere.sdk.features.TTS.Services.AudioPlaybackManager
import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.isActive
import kotlin.coroutines.cancellation.CancellationException

/**
 * Captures mic audio and feeds raw frames to the in-core voice agent bound to
 * [handle]. [run] suspends until the calling coroutine is cancelled; cancel it
 * to stop the session (capture teardown is handled in a finally block).
 *
 * Segmentation/endpointing lives in the C core, which re-runs its own VAD over
 * each utterance and is strictly turn-taking (no barge-in). Mic frames that
 * arrive while a turn is processing are dropped by the bounded channel, which
 * also avoids transcribing the device's own TTS output.
 */
internal class VoiceAgentMicDriver(
    private val handle: Long,
) {
    private val logger = SDKLogger("VoiceAgentMic")
    private val capture = AudioCaptureManager()
    private val playback = AudioPlaybackManager()

    suspend fun run() {
        val chunks =
            Channel<ByteArray>(
                capacity = MIC_CHANNEL_CAPACITY,
                onBufferOverflow = BufferOverflow.DROP_OLDEST,
            )
        capture.startRecording { chunk -> chunks.trySend(chunk) }
        logger.info("Voice-agent mic capture started")
        try {
            feedLoop(chunks)
        } finally {
            capture.stopRecording()
            playback.stop()
            chunks.close()
            logger.info("Voice-agent mic capture stopped")
        }
    }

    private suspend fun feedLoop(chunks: Channel<ByteArray>) {
        while (currentCoroutineContext().isActive) {
            val chunk = chunks.receive()

            val resultBytes =
                try {
                    RunAnywhereBridge.racVoiceAgentFeedAudioProto(
                        handle,
                        chunk,
                        SAMPLE_RATE_HZ,
                        1,
                        AudioEncoding.AUDIO_ENCODING_PCM_S16_LE.value,
                        false,
                    )
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Throwable) {
                    // Never swallow JVM Errors (OOM, …). A recoverable failure
                    // means this utterance's turn failed (e.g. empty STT) or the
                    // agent was torn down; the session cancels this coroutine on
                    // teardown, so log and keep feeding rather than killing the
                    // loop on a single bad turn.
                    if (e is Error) throw e
                    logger.warning("Voice feed failed: ${e.message}")
                    null
                } ?: continue

            val result =
                try {
                    VoiceAgentResult.ADAPTER.decode(resultBytes)
                } catch (_: Exception) {
                    null
                } ?: continue

            // A non-empty reply means the core closed an utterance and ran a full
            // turn this call. synthesized_audio is self-describing WAV.
            val reply = result.synthesized_audio
            if (reply != null && reply.size > 0) {
                logger.info("Playing agent reply (${reply.size} WAV bytes)")
                playReply(reply.toByteArray())
                // Drop frames captured while the turn ran / the device spoke so
                // stale audio is not folded into the next turn.
                while (chunks.tryReceive().isSuccess) Unit
            }
        }
    }

    private suspend fun playReply(wav: ByteArray) {
        if (wav.isEmpty()) return
        try {
            playback.play(wav)
        } catch (e: CancellationException) {
            playback.stop()
            throw e
        } catch (e: Exception) {
            logger.warning("Agent reply playback failed: ${e.message}")
        }
    }

    private companion object {
        const val SAMPLE_RATE_HZ = 16_000

        /**
         * Bounded mic ingress buffer. The capture callback trySends while the
         * consumer pauses for the duration of each turn, so an unbounded channel
         * could grow without limit on long turns. DROP_OLDEST bounds memory;
         * frames captured mid-turn are discarded anyway (no barge-in).
         */
        const val MIC_CHANNEL_CAPACITY = 128
    }
}
